crates/noirc_evaluator/src/ssa/opt/flatten_cfg.rs

```diff
@@ -245,6 +259,26 @@ impl<'f> Context<'f> {
     /// Returns the last block to be inlined. This is either the return block of the function or,
     /// if self.conditions is not empty, the end block of the most recent condition.
     fn handle_terminator(&mut self, block: BasicBlockId) -> BasicBlockId {
+        if let TerminatorInstruction::JmpIf { .. } =
+            self.inserter.function.dfg[block].unwrap_terminator()
+        {
+            // Find stores in the outer block and insert into the `outer_block_stores` map.
+            // Not using this map can lead to issues when attempting to merge slices.
+            // When inlining a branch end, only the then branch and the else branch are checked for stores.
+            // However, there are cases where we want to load a value that comes from the outer block
+            // that we are handling the terminator for here.
```

Comment:
I'm confused why this would be needed now with slices and not before with other values. 

> // When inlining a branch end, only the then branch and the else branch are checked for stores.
   // However, there are cases where we want to load a value that comes from the outer block
 
I would think `store_values` should still be able to handle cases where only one block is stored to, with the other inheriting from the parent value. What cases break this?

Reply:
With slices we _need_ to fetch the length from the current value or previously stored value. With arrays, the length is already inherently part of the type. 

For example in a program like such: 
```
fn dynamic_slice_index_if(mut slice: [Field], x: Field) {
    if x as u32 < 10 {
        assert(slice[x] == 4);
        slice[x] = slice[x] - 2;
    } else {
        assert(slice[x] == 0);
    }
    assert(slice[4] == 2);
}
``` 
We have this SSA before flattening:
```
After Mem2Reg:
acir fn main f0 {
  b0(v0: Field):
    v128 = sub v0, Field 1
    v129 = sub v0, Field 4
    v130 = allocate
    v131 = allocate
    store Field 5 at v130
    store [Field 0, Field 1, Field 2, Field 3, Field 4] at v131
    v162 = allocate
    store Field 5 at v162
    v163 = allocate
    store [Field 0, Field 1, Field 2, Field 3, Field 4] at v163
    v165 = cast v128 as u32
    v166 = lt v165, u32 10
    jmpif v166 then: b4, else: b5
  b4():
    v171 = lt v128, u64 5
    constrain v171 == u1 1
    v173 = array_get [Field 0, Field 1, Field 2, Field 3, Field 4], index v128
    constrain v173 == Field 4
    v174 = lt v128, u64 5
    constrain v174 == u1 1
    v176 = array_get [Field 0, Field 1, Field 2, Field 3, Field 4], index v128
    v177 = sub v176, Field 2
    v179 = array_set [Field 0, Field 1, Field 2, Field 3, Field 4], index v128, value v177, length Field 5
    v180 = add v128, Field 1
    store Field 5 at v162
    store v179 at v163
    jmp b6()
  b6():
    v181 = load v163
    v182 = array_get v181, index Field 4
    constrain v182 == Field 2
    return 
  b5():
    v168 = lt v128, u64 5
    constrain v168 == u1 1
    v170 = array_get [Field 0, Field 1, Field 2, Field 3, Field 4], index v128
    constrain v170 == Field 0
    jmp b6()
}
```
If the input to `dynamic_slice_index_if` was `[Field; 5]` the SSA would look almost exactly the same just without the length checks before every `array_get` call. As you can see there is no store in `b5` so it is unclear what value is stored in `v163` if only the then branch and else branch are checked for stores. We must check the parent block to see whether there has been a store whose value can tell us the slice length. For an array merger, we already have the length in the type so we don't have to go searching for the original value.

Reply:
`store_values` is only inserted into when inlining a branch. If the store is in the branch being inlined into, `store_values` is not going to have the value.